File: python/bowling/bowling.py
"""Score a bowling card."""
from __future__ import annotations


class Frame:
    """One frame of a game."""

    next_frame: Frame

    def __init__(self, frame: int):
        """Initialize."""
        self.rolls: list[int] = []
        self.frame = frame

    @property
    def is_open_frame(self) -> bool:
        """Return if this frame is open."""
        return sum(self.rolls) < 10

    @property
    def is_spare(self) -> bool:
        """Return if this frame contains a spare."""
        return self.rolls[0] != 10 and sum(self.rolls[:2]) == 10

    @property
    def is_strike(self) -> bool:
        """Return if this frame contains a strike."""
        return 10 in self.rolls

    def roll(self, pins: int) -> None:
        """Validate and track a roll."""
        if self.frame > 10:
            raise ValueError("game is over")
        if not 0 <= pins <= 10:
            raise ValueError(f"invalid pin value, {pins}")

        if pins == 10 and self.rolls:
            # Strike is only possible on the first roll of a frame
            # ... or on frame 10 (following a strike/spare).
            assert self.frame == 10 and sum(self.rolls) >= 10, "invalid pin value"

        # Combined pins on the second* roll cannot exceed 10.
        # On frame 10, unless following a spare/strike, combined pins cannot exceed 10.
        if self.rolls:
            if self.frame != 10:
                assert sum(self.rolls) + pins <= 10, "invalid pin value"
            elif self.rolls[-1] != 10 and not self.is_spare:
                assert self.rolls[-1] + pins <= 10, "invalid pin value"

        self.rolls.append(pins)

    def full(self) -> bool:
        """Return if this frame is "full" or complete."""
        # Normally, a frame is over on the second ball or a strike.
        if self.frame < 10:
            return self.is_strike or len(self.rolls) == 2

        # On the last frame, you get a third roll when you have a spare/strike.
        if self.is_spare or self.is_strike:
            return len(self.rolls) == 3
        return len(self.rolls) == 2

    def get_next(self) -> Frame:
        """Return a new next frame."""
        self.next_frame = Frame(self.frame + 1)
        return self.next_frame

    def next_two_rolls(self) -> list[int]:
        """Return the next two rolls after this frame."""
        rolls = self.next_frame.rolls[:2]
        if len(rolls) == 1:
            rolls.append(self.next_frame.next_frame.rolls[0])
        return rolls

    def score(self) -> int:
        """Return the score of a frame."""
        if self.is_open_frame:
            return sum(self.rolls)
        if self.is_spare:
            return sum(self.rolls + self.next_two_rolls()[:1])
        if self.is_strike:
            return sum(self.rolls + self.next_two_rolls()[:2])
        raise RuntimeError("Invalid frame")

    def __str__(self) -> str:
        """Return a string reresentation of a frame."""
        return f"{self.frame}{self.rolls}"


class BowlingGame:
    """Bowling score."""

    def __init__(self):
        """Initialize."""
        self.rolls: list[int] = []
        self.current_frame = Frame(1)
        self.frames: list[Frame] = [self.current_frame]

    def roll(self, pins: int) -> None:
        """Validate and track a roll."""
        self.current_frame.roll(pins)
        if self.current_frame.full():
            self.current_frame = self.current_frame.get_next()
            self.frames.append(self.current_frame)

    def score(self) -> int:
        """Score a game."""
        assert self.current_frame.frame == 11, "game not over"
        return sum(f.score() for f in self.frames)

File: go/bowling/bowling.go
package bowling

import (
	"errors"
	"fmt"
	"slices"
)

// sum returns the sum of a slice of ints.
func sum(vals []int) int {
	var total int
	for _, v := range vals {
		total += v
	}
	return total
}

// Frame stores one game frame.
type Frame struct {
	num   int
	rolls []int
	next  *Frame
}

// nextTwoRolls returns two more rolls for use in computing a strike or spare.
func (f *Frame) nextTwoRolls() []int {
	rolls := slices.Clone(f.next.rolls)
	if len(rolls) > 2 {
		rolls = rolls[:2]
	} else if len(rolls) < 2 {
		rolls = append(rolls, f.next.next.rolls[0])
	}
	return rolls
}

// score returns the score of a frame, handling a strike or spare.
func (f *Frame) score() int {
	if len(f.rolls) == 0 {
		return 0
	}
	total := sum(f.rolls)
	if f.num < 9 {
		if f.rolls[0] == 10 {
			total += sum(f.nextTwoRolls())
		} else if total == 10 {
			total += f.nextTwoRolls()[0]
		}
	}
	return total
}

// full returns if a frame is full/complete.
func (f *Frame) full() bool {
	if f.num < 9 {
		return len(f.rolls) == 2 || (len(f.rolls) == 1 && f.rolls[0] == 10)
	}
	return len(f.rolls) == 3 || (len(f.rolls) == 2 && f.rolls[0]+f.rolls[1] < 10)
}

// Game represents a bowling game.
type Game struct {
	frames   []*Frame
	frameIdx int
}

// NewGame returns a new Game with the first frame ready to use.
func NewGame() *Game {
	return &Game{frames: []*Frame{&Frame{}}}
}

// bumpFrame adds the next Frame to a game if the current frame is full.
func (g *Game) bumpFrame() {
	frame := g.frames[g.frameIdx]
	if frame.full() && g.frameIdx < 10 {
		g.frameIdx++
		new := &Frame{num: g.frameIdx}
		frame.next = new
		g.frames = append(g.frames, new)
	}
}

// Roll adds a roll to the game.
func (g *Game) Roll(pins int) error {
	if pins < 0 || pins > 10 {
		return errors.New("invalid pins")
	}
	frame := g.frames[g.frameIdx]
	var priorRoll int
	if len(frame.rolls) > 0 {
		priorRoll = frame.rolls[len(frame.rolls)-1]
	}
	if pins == 60 {
		fmt.Println("prior", priorRoll)
	}
	if g.frameIdx < 9 && len(frame.rolls) > 0 && priorRoll != 10 && priorRoll+pins > 10 {
		return errors.New("invalid frame")
	}
	if g.frameIdx == 9 && sum(frame.rolls) != 10 && priorRoll != 10 && priorRoll+pins > 10 {
		return errors.New("invalid frame")
	}
	if g.frameIdx >= 10 {
		return errors.New("invalid game over")
	}
	frame.rolls = append(frame.rolls, pins)
	g.bumpFrame()

	return nil
}

// Score returns the score of a game.
func (g *Game) Score() (int, error) {
	if g.frameIdx != 10 {
		return 0, errors.New("game not over")
	}
	var total int
	for _, f := range g.frames {
		total += f.score()
	}
	return total, nil
}
